File: LeetcodeSolutions/Hashing/GroupAnagrams.java
import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Arrays;
public class GroupAnagrams {
    public static void main(String[] args) {
        String[] strs = new String[] {"eat", "tea", "tan", "ate", "nat", "bat"};
        System.out.println(groupAnagrams(strs).toString());
    }

    public static List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> result = new ArrayList<>();
        if (strs.length == 0) return result;

        HashMap<String, List<String>> map = new HashMap<>();

        for (String s : strs) {
            int[] hash = new int[26];
            for (char c: s.toCharArray()) {
                hash[c - 'a']++;
            }
            String key = Arrays.toString(hash);
            map.computeIfAbsent(key, k -> new ArrayList<String>());
            map.get(key).add(s);
        }
        result.addAll(map.values());
        return result;
    }


}

File: LeetcodeSolutions/Hashing/GroupAnagrams.py
from collections import defaultdict;
def groupAnagrams(strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        if len(strs) == 0: 
            return []
        
        result = defaultdict(list)
        for str in strs:
            word = [0] * 26
            for c in str:
                word[ord(c) - ord("a")] += 1
            result[tuple(word)].append(str)
        
        return result.values()

def groupAnagrams2(strs):
     """
    :type strs: List[str]
    :rtype: List[List[str]]
    """
     anagram_map = defaultdict(list)

     for str in strs:
          sorted_word = ''.join(sorted(str))
          anagram_map[sorted_word].append(str)
     return anagram_map.values()

strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
print(groupAnagrams(strs))
print(groupAnagrams2(strs))

     